lib/SILOptimizer/Transforms/AccessEnforcementDom.cpp

```diff
@@ -65,417 +57,549 @@
-  if (!beginAccess->hasNoNestedConflict()) {
-    return;
-  }
+protected:
+  void setBottom(BBState &state);
+  void analyzeAccess(BeginAccessInst *BAI, BBState &state);
+};
+} // namespace
 
-  // check that the storage we found is identical to the
-  // one we found in order to enable optimization
-  auto parentStorage = visitedAccessesIt->second;
-  if (!parentStorage.hasIdenticalBase(storage)) {
+// Set information for all in-scope accesses to the worst case "bottom".
+// The isInner flag is not affected because it only applies to scopes inside the
+// currently open scopes, not the currently open scopes themselves.
+void DominatedAccessAnalysis::setBottom(BBState &state) {
+  if (state.isBottom)
     return;
-  }
 
-  auto *parentBegin = visitedAccessesIt->first;
+  // Unordered iteration over the in-access scopes.
+  llvm::for_each(state.inScopeAccesses, [this](BeginAccessInst *BAI) {
+    if (auto &domStorage = result.accessMap[BAI])
+      domStorage.setContainsRead();
+  });
+}
 
-  // If this beginAccess is a modify, and is nested inside parent
-  // then we have to bail: can't change it to static
-  // If, on the other hand, it is a read, we can continue.
-  if (beginAccess->getAccessKind() >= SILAccessKind::Modify) {
-    auto *endP = getSingleEndAccess(parentBegin);
-    if (!endP) {
-      return;
-    }
-    if (!domInfo->properlyDominates(endP, beginAccess)) {
-      return;
+// Perform the analysis and return the Result, relinquishing internal state.
+DominatedAccessAnalysis::Result DominatedAccessAnalysis::analyze() && {
+  // A single RPO traversal is sufficient to visit access scopes in order. The
+  // set of open accesses entering a block cannot grow as a result of loops, and
+  // within an open scope the results are flow-insensitive.
+  for (auto *BB : PO->getReversePostOrder()) {
+    BBState state = blockStateMap[BB];
+    for (auto &I : *BB) {
+      if (auto *BAI = dyn_cast<BeginAccessInst>(&I)) {
+        analyzeAccess(BAI, state);
+        continue;
+      }
+      if (auto *EAI = dyn_cast<EndAccessInst>(&I)) {
+        bool erased = state.inScopeAccesses.erase(EAI->getBeginAccess());
+        (void)erased;
+        assert(erased);
+        continue;
+      }
+      // Check for BeginApply before checking FullApplySite below.
+      if (auto *beginApply = dyn_cast<BeginApplyInst>(&I)) {
+        auto iterAndInserted = state.inScopeCoroutines.insert(beginApply);
+        (void)iterAndInserted;
+        assert(iterAndInserted.second);
+        continue;
+      }
+      if (auto *endApply = dyn_cast<EndApplyInst>(&I)) {
+        bool erased = state.inScopeCoroutines.erase(endApply->getBeginApply());
+        (void)erased;
+        assert(erased);
+        continue;
+      }
+      if (FullApplySite::isa(&I)) {
+        setBottom(state);
+        continue;
+      }
+      if (isa<BeginUnpairedAccessInst>(&I)) {
+        // Unpaired accesses could be tracked, but are ignored because they are
```

Comment:
I am not sure this is 100% safe, but as you said - I concur: being too conservative is an overkill

Reply:
Oops, my comment was misleading; I fixed it:

        // Unpaired accesses could be tracked, but are ignored because they are
        // mostly irrelevant and hard to test. Completely bail on this function.

We can never hit this in practice. The only time an unpaired access is ever created is in non-inlinable code in the stdlib via a builtin. I suspect we could get rid of the instruction, but I would need to think harder about proving the KeyPath code would be safe without it.